### DISCUSSION TITLE
A2W gained weight after using Groundwork

### DISCUSSION THREAD

**Discussion by adamscarberry** [USACE/groundwork#64] [General]
The lighthouse performance score went from 76 to 56.

Trying to visualize the assets load below using https://www.npmjs.com/package/rollup-plugin-visualizer

<img width="2205" alt="image" src="https://github.com/USACE/groundwork/assets/46494851/4ebae7ce-f73e-4c81-a109-83f8a85435a4">

This is the build output:

<img width="747" alt="image" src="https://github.com/USACE/groundwork/assets/46494851/1a757f5f-83d3-48d0-82a7-1d5fe43a644c">

---
Also I had to give node more memory in AWS code build, otherwise it ran out of memory.

I suspect if we removed the Water Management components/hooks/etc, we might see a slim down. Maybe they should go in a separate npm package?

cc @willbreitkreutz

**Comment by adamscarberry** [comment]
I ran the same tool on the groundwork lib:

<img width="1902" alt="image" src="https://github.com/USACE/groundwork/assets/46494851/378f78e4-23a6-48e2-8fc8-a7cad6fb6d83">

**Comment by adamscarberry** [comment]
I ripped out plotly locally only just to see what would be left:

<img width="1646" alt="image" src="https://github.com/USACE/groundwork/assets/46494851/629effd3-581b-47d2-90fc-4e0f4ef8a363">

<img width="457" alt="image" src="https://github.com/USACE/groundwork/assets/46494851/2bfc0cd6-826f-46fb-b4a4-6870c79ae723">

---

So it looks like plotly is the pig.

**Comment by willbreitkreutz** [comment]
@adamscarberry thanks for looking at that, it doesn't surprise me that plotly would be the pig.  Might break out `groundwork-water` `groundwork-plots` with the existing `groundwork-geo`

**Comment by krowvin** [comment]
### Throwing my $0.02 in the ring
> Maybe they should go in a separate npm package  

> groundwork-water  

The timing is wild here. 
I have been thinking district specific components might be better suited in a mono repo [here](https://github.com/USACE-WaterManagement).  
The `../usace-watermanagement` repo. 

And then any components that would be maintained by GroundWork could be bubbled up to that level. 
This makes it easier for any given district to maintain their _own_ components. 

Alternatively, if groundwork wanted to use those templates it could call them as dependencies instead of assuming control (and maintenance) of them. 

Then as you mention  
> a separate npm package  

Either per district or for the whole repo. I.e. a sep NPM project within each directory of a mono repo. 

## For example
`/usace-watermanagement/groundwork/swt/components/DroughtMap.jsx`  
`/usace-watermanagement/groundwork/swt/package.json`  

### And summoning a few others for this - as it seems like a larger decision 
@MikeNeilson  @DanielTOsborne @jbkolze  @JeremyDKellett 

#### Edit:
`/usace-watermanagement/groundwork` could be confusing. 
Perhaps `groundwork-water` as Will mentions, but in this ~repo~ org.

**Comment by willbreitkreutz** [comment]
I'm curious what others think, but I would caution against flaking out into lots of district based packages.  I think `groundwork-water` makes sense, any sharable components districts develop should be contributed there, any more generic components would be contributed to `groundwork` proper.  Any components that are specific enough that they would only be used by one district would just live in that separate consumer application.

**Comment by MikeNeilson** [comment]
I'm going to, i think, agree with Will here. While there is certainly a separation of concerns - not all districts water management mission is the same - but there are similarities, it likely makes more sense to just organize by concept vs district.

Maybe with some exceptions for some real doozy edge cases. (e.g. no one should do the rounding LRE is required to do "by law", except LRE of course.).

**Comment by MikeNeilson** [comment]
This is also starting new, the projects were I'm pushing "here's a separate place for district stuff" is because I'm pretty sure it'll be an easier transition.

**Comment by jbkolze** [comment]
If we're separating repos, I think it probably does make sense that groundwork-water be hosted within the WM org.  I see base groundwork as the foundational design/layout components, with any offshoots of that applying the design for org-specific purposes / components.  Seems to follow that the org would be the owners/stewards of that part of it.

I do think it will be important to ensure that layout-focused changes/enhancements that *aren't* org-specific get rolled up into base groundwork.  It will be a lot easier to lose sight of that if repos are separate, and one of the main points is to keep design aspects as similar across-the-board as is feasible.

Agree that managing all district-specific components in one repo could get messy, and that those are probably best stored in their dependent applications (or potentially in a district-owned repo if used in a variety of apps).

**Comment by krowvin** [comment]
> I'm curious what others think, but I would caution against flaking out into lots of district based packages. I think `groundwork-water` makes sense, any sharable components districts develop should be contributed there, any more generic components would be contributed to `groundwork` proper. Any components that are specific enough that they would only be used by one district would just live in that separate consumer application.

Well this makes sense to me! I think Will irons it out well here too. 

__So Then__ If I follow:
* If it's generic it goes in `usace/groundwork`  
* If it is WM but not district specific it goes in `usace-watermanagement/groundwork-water`  
* If it is WM and district specific it goes in a given district's org (ideally)?

**Reply by krowvin** [reply]
Heads up, a repo was created here for item 2 in this list
https://github.com/USACE-WaterManagement/groundwork-water

**Reply by adamscarberry** [reply]
@krowvin @willbreitkreutz Maybe we add links to the groundwork spinoffs in the groundwork readme.  Help folks who are looking for WM components and direct them to the new repo?

**Comment by DanielTOsborne** [comment]
Forgive me, as I'm not a web developer, so I don't know how any of this really works.

However, I was under the impression React would strip out what is not being used.
Why would plotly even be in the final output if A2W isn't using it?

The version of plotly I use is a custom build I made to strip out components I'm not using and reduce the file size (and fix some bugs not fixed upstream). However I'm just using the dist output directly in raw javascript, not through a framework.
Probably not an option for general use though.

**Reply by willbreitkreutz** [reply]
The build tooling should be stripping out any unused code (tree-shaking), I think there is some re-organization of imports/exports that could solve the issue, something I need to dig into more.

**Comment by adamscarberry** [comment]
@willbreitkreutz 

Either of these issues seem related or make sense to you?

https://github.com/vitejs/vite/issues/7970

https://github.com/vitejs/vite/issues/5676

**Reply by willbreitkreutz** [reply]
possibly... but I think the long-term solution is to pull charting out into a separate lib, or possibly even better, just write documentation on how to include it from a direct install of plotly and not make it part of groundwork
